src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -318,129 +370,117 @@ def _post_process(
-                host_assembly_dir,
-                assembly_name,
-                upload_dir,
-                conf.output_dir,
-                upa
-            )
-        except Exception as e:
-            print(f"Failed assembly name: {assembly_name}. Exception:")
-            print(e)
-
-        conf.output_queue.put((assembly_name, upa))
-
-        if counter % 3000 == 0:
-            print(f"Assemblies processed: {counter}/{assembly_files_len}, "
-                  f"Percentage: {counter / assembly_files_len * 100:.2f}%, "
-                  f"Time: {datetime.now()}")
-
-
 def _upload_assembly_files_in_parallel(
-        conf: Conf,
+        asu: AssemblyUtil,
+        ws: Workspace,
         upload_env_key: str,
         workspace_id: int,
+        load_version: str,
         upload_dir: str,
         wait_to_upload_assemblies: dict[str, str],
+        batch_size: int,
+        output_dir: str,
 ) -> list[str]:
     """
     Upload assembly files to the target workspace in parallel using multiprocessing.
 
     Parameters:
-        conf: Conf object
+        asu: AssemblyUtil client
+        ws: Workspace client
         upload_env_key: environment variable key in uploaded.yaml file
         workspace_id: target workspace id
+        load_version: load version
         upload_dir: a directory in collectionssource that creates new directories linking to sourcedata
         wait_to_upload_assemblies: a dictionary that maps assembly file name to assembly directory
+        batch_size: a number of files to upload per batch
+        output_dir: a directory in sourcedata/workspace to store new assembly entries
 
     Returns:
-        a list of assembly names that failed to upload
+        number of assembly files have been sucessfully uploaded from wait_to_upload_assemblies
     """
     assembly_files_len = len(wait_to_upload_assemblies)
-    print(f"Start uploading {assembly_files_len} assembly files with {conf.workers} workers\n")
+    print(f"Start uploading {assembly_files_len} assembly files\n")
 
-    # Put the assembly files in the input_queue
-    counter = 1
-    for assembly_name, host_assembly_dir in wait_to_upload_assemblies.items():
+    uploaded_count = 0
+    uploaded_fail = False
+    for assembly_tuples in _gen(wait_to_upload_assemblies, batch_size):
 
-        container_internal_assembly_path = os.path.join(
-            JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, assembly_name
-        )
-        conf.input_queue.put(
-            (
+        try:
+            batch_upas = _upload_assemblies_to_workspace(asu, workspace_id, assembly_tuples)
+            batch_uploaded_tuples = assembly_tuples
+        except Exception as e:
+            print(e)
+            name2tuple = {assembly_tuple.assembly_name: assembly_tuple for assembly_tuple in assembly_tuples}
+            refs = [{"wsid": workspace_id, "name": name} for name in name2tuple]
+            res = ws.get_object_info3({"objects": refs, "ignoreErrors": 1})
```

Comment:
If this call fails then the script will exit with data inconsistencies between the workspace and the yaml files, so the user should be warned about that. 

Also, this means the effective max batch size is 10k - is that enforced anywhere?

Reply:
ah, does get_object_info3 have a limit of 10000? I thought only list_objects func has this limitation?

Reply:
Yep. Most workspace calls have limits, otherwise you'd get people asking for hundreds of thousands of objects and monopolizing or crashing the server

Reply:
What arguments I should pass in to limit get_object_info3? 
I cannot find it here: https://kbase.us/services/ws/docs/Workspace.html#funcdefWorkspace.get_object_info3

Reply:
None, you just ask for no more than 10k objects

Reply:
Now I see what you are saying.

Reply:
👍

Reply:
:+1: